crates/revm/src/journaled_state.rs

```diff
@@ -468,39 +459,55 @@ impl JournaledState {
+    /// current spec enables Cancun, this happens only when the account associated to [address]
+    /// is created in the same tx
+    /// references:
+    ///  * https://github.com/ethereum/go-ethereum/blob/141cd425310b503c5678e674a8c3872cf46b7086/core/vm/instructions.go#L832-L833
+    ///  * https://github.com/ethereum/go-ethereum/blob/141cd425310b503c5678e674a8c3872cf46b7086/core/state/statedb.go#L449
+    ///  * https://eips.ethereum.org/EIPS/eip-6780
     pub fn selfdestruct<DB: Database>(
         &mut self,
         address: B160,
         target: B160,
         db: &mut DB,
     ) -> Result<SelfDestructResult, DB::Error> {
         let (is_cold, target_exists) = self.load_account_exist(target, db)?;
-        // transfer all the balance
-        let acc = self.state.get_mut(&address).unwrap();
-        let balance = mem::take(&mut acc.info.balance);
-        let previously_destroyed = acc.is_selfdestructed();
-        acc.mark_selfdestruct();
-
-        // NOTE: In case that target and destroyed addresses are same, balance will be lost.
-        // ref: https://github.com/ethereum/go-ethereum/blob/141cd425310b503c5678e674a8c3872cf46b7086/core/vm/instructions.go#L832-L833
-        // https://github.com/ethereum/go-ethereum/blob/141cd425310b503c5678e674a8c3872cf46b7086/core/state/statedb.go#L449
-        if address != target {
-            let target_account = self.state.get_mut(&target).unwrap();
-            // touch target account
+
+        let acc = if address != target {
+            let [acc, target_account] = self.state.get_many_mut([&address, &target]).unwrap();
             Self::touch_account(self.journal.last_mut().unwrap(), &target, target_account);
-            target_account.info.balance += balance;
-        }
+            target_account.info.balance += acc.info.balance;
+            acc
+        } else {
+            self.state.get_mut(&address).unwrap()
+        };
 
-        self.journal
-            .last_mut()
-            .unwrap()
-            .push(JournalEntry::AccountDestroyed {
+        let balance = acc.info.balance;
+        let previously_destroyed = acc.is_selfdestructed();
+        let is_cancun_enabled = SpecId::enabled(self.spec, CANCUN);
+
+        // EIP-6780 (Cancun hard-fork): selfdestruct only if contract is created in the same tx
+        let journal_entry = if acc.is_created() || !is_cancun_enabled {
+            acc.mark_selfdestruct();
+            acc.info.balance = U256::ZERO;
+            JournalEntry::AccountDestroyed {
                 address,
                 target,
                 was_destroyed: previously_destroyed,
                 had_balance: balance,
-            });
+            }
+        } else {
+            Self::touch_account(self.journal.last_mut().unwrap(), &address, acc);
+            JournalEntry::BalanceTransfer {
+                from: address,
+                to: target,
+                balance,
+            }
```

Comment:
We are not decrementing the balance from `address`

Reply:
Oof what a dumb error, thanks for noticing